components/style/rule_tree/core.rs

```diff
@@ -119,16 +84,18 @@ impl RuleTree {
     }
 
     /// This can only be called when no other threads is accessing this tree.
-    pub unsafe fn gc(&self) {
-        self.root.gc();
+    pub fn gc(&self) {
+        unsafe { self.swap_free_list_and_gc(RuleNode::DANGLING_PTR) }
     }
 
     /// This can only be called when no other threads is accessing this tree.
-    pub unsafe fn maybe_gc(&self) {
+    pub fn maybe_gc(&self) {
         #[cfg(debug_assertions)]
         self.maybe_dump_stats();
 
-        self.root.maybe_gc();
+        if self.root.p.free_count.load(Ordering::Relaxed) > RULE_TREE_GC_INTERVAL {
```

Comment:
What increments the free count now? Can this ever trigger?

Reply:
Added the incrementing back, hah.